### PR TITLE
feat: AirGap Harvester With Rancher

### PR DESCRIPTION
**Pull request by irishgordo**
* provides provisioning for airgapping Rancher with Harvester via the ipxe vagrant workflow

Resolves: feat/rancher-and-harvester-airgap-provisioning

This is the newer PR, based on:
https://github.com/harvester/ipxe-examples/pull/45

### CODE REVIEW

**Comment by irishgordo** [comment]
Looking into timeout issue, acquiring the rancher-version.tgz package

**Comment by irishgordo** [comment]
Need to add system settings for registry:
https://docs.harvesterhci.io/v1.1/advanced/settings/#containerd-registry

**Comment by TachunLin** [comment]
Then I try importing Harvester to Rancher, 

* Before we import the Harvester Cluster into the Rancher VM, we need to modify the containerd registry in Harvester setting  
image


```
{
  "Mirrors": {
    "docker.io": {
      "Endpoints": [
        "myregistry.local:5000"
      ],
      "Rewrites": null
    },
    "myregistry.local": {
      "Endpoints": [
        "myregistry.local:5000"
      ],
      "Rewrites": null
    },
    "myregistry.local:5000": {
      "Endpoints": [
        "myregistry.local:5000"
      ],
      "Rewrites": null
    }
  },
  "Configs": {
    "myregistry.local:5000": {
      "Auth": null,
      "TLS": {
        "CAFile": "",
        "CertFile": "",
        "KeyFile": "",
        "InsecureSkipVerify": true
      }
    }
  },
  "Auths": null
}
```

* But on Rancher virtualization management page, the Harvester cluster remain in the `Pending` state
image

* I try adding host record in the harvester-node-0 /etc/hosts file , but seems not work
```
192.168.2.34    rancher-vagrant-vm.local myregistry.local
```

Not sure whether I miss any steps or set incorrect containerd settings

**Comment by irishgordo** [comment]
> Then I try importing Harvester to Rancher,
> 
>     * Before we import the Harvester Cluster into the Rancher VM, we need to modify the containerd registry in Harvester setting
>       image
> 
> 
> ```
> {
>   "Mirrors": {
>     "docker.io": {
>       "Endpoints": [
>         "myregistry.local:5000"
>       ],
>       "Rewrites": null
>     },
>     "myregistry.local": {
>       "Endpoints": [
>         "myregistry.local:5000"
>       ],
>       "Rewrites": null
>     },
>     "myregistry.local:5000": {
>       "Endpoints": [
>         "myregistry.local:5000"
>       ],
>       "Rewrites": null
>     }
>   },
>   "Configs": {
>     "myregistry.local:5000": {
>       "Auth": null,
>       "TLS": {
>         "CAFile": "",
>         "CertFile": "",
>         "KeyFile": "",
>         "InsecureSkipVerify": true
>       }
>     }
>   },
>   "Auths": null
> }
> ```
> 
>     * But on Rancher virtualization management page, the Harvester cluster remain in the `Pending` state
>       image
> 
>     * I try adding host record in the harvester-node-0 /etc/hosts file , but seems not work
> 
> 
> ```
> 192.168.2.34    rancher-vagrant-vm.local myregistry.local
> ```
> 
> Not sure whether I miss any steps or set incorrect containerd settings

@TachunLin Thank you for taking a look at this :+1: :smile: 

What does it look like when you navigate to the Embedded Rancher UI in Harvester?
Are there any noticeable errors surrounding `cattle-cluster-agent` it's replicaset, deployment & pods in the `cattle-system` namespace?
Screenshot from 2023-03-20 11-55-32
Wondering if something happened there :thinking: 
At: https://192.168.2.131/dashboard/c/local/explorer#cluster-metrics

You shouldn't have to edit the harvester-node-0 /etc/hosts file.
Also I've not ever commented out the:
```
libvirt.loader = '/usr/share/qemu/OVMF.fd'
```
Line in the Vagrantfile.  Is there a reason why we would want the legacy BIOS for this testing?

I'm guessing the pending status on Rancher when importing, seems like it could be a registry issue / it being able to pull the correct image of rancher-agent.
When you navigate to:
https://192.168.2.34:5000/v2/_catalog?n=500
Do you see the `rancher-agent` image present in the registry?
Thinking this might just be registry configuration issue.
Perhaps would need to scale the replicas to 0 for the deployment then scale back up to 2 depending on the issue / when the containerd registry edits were made.

**Comment by TachunLin** [comment]
@irishgordo , Thank you for checking and the troubleshooting suggestion
 
While navigating to the cluster-metrics of https://192.168.2.131/dashboard/c/local/explorer#cluster-metrics
I encounter the ImagePullBackOff error and looping to pull image `rancher-agent:2.6.11`
  image

And try navigating to https://192.168.2.34:5000/v2/_catalog?n=500 but can't reach the site
image


The reason to comment out the line is without this action, the provisioning of Harvester node would got provisioning failure on the KVM libvirt of my current OpenSUSE Leap 15.3 desktop. 
```
libvirt.loader = '/usr/share/qemu/OVMF.fd'
```

**Comment by irishgordo** [comment]
@TachunLin  thanks for providing more of that info :smile: :+1: 

I think I see a few things :smile:

Thinking perhaps navigating to:
https://myregistry.local:5000/v2/_catalog?n=500
Would take result in the catalog being able to be displayed.
As the `rancher-agent` should be present, in that list.
With:
```
╭─mike at suse-workstation-team-harvester in ~/Projects/v112-rc3-testing/v1.1 on main✘✘✘
╰─± cat /etc/hosts
127.0.0.1	localhost
::1		localhost
127.0.1.1	pop-os.localdomain	pop-os
192.168.2.34	rancher-vagrant-vm.local	myregistry.local
```
( the edit of myregistry.local , for the docker-registry in the etc/hosts )

I'm also thinking that the removal of the additional mirrors would be beneficial in the containerd registry setting in harvester, so just having:
```json
{
  "Mirrors": {
    "docker.io": {
      "Endpoints": [
        "myregistry.local:5000"
      ],
      "Rewrites": null
    }
  },
  "Configs": {
    "myregistry.local:5000": {
      "Auth": null,
      "TLS": {
        "CAFile": "",
        "CertFile": "",
        "KeyFile": "",
        "InsecureSkipVerify": true
      }
    }
  },
  "Auths": null
}
```

In, https://192.168.2.131/dashboard/c/local/explorer/apps.deployment - with "All Namespaces" selected in dropdown:
Screenshot from 2023-03-22 14-15-54
Then, after making that edit to the containerd registry to either:
- delete the deployment, remove the virtualization in Rancher & create a new one
- or try to edit the replicas to 0, then save, then edit replicas to 2

If there isn't any activity of it trying to pull the rancher-agent from the docker-registry.

**Comment by TachunLin** [comment]
@irishgordo , thank you for the checking and instruction. 
I try navigating to https://myregistry.local:5000/v2/_catalog?n=500 but still display page not found

Check my host machine host file in /etc/hosts, the related record have added. 
```
127.0.0.1       localhost
192.168.2.34    rancher-vagrant-vm.local myregistry.local
```

Then update the containerd registry setting to remove additional mirrors 
image

After making the edit, delete the `Updating` cattle-cluser-agent:rancher/rancher-agent:v2.6.11
image

Remove the import in Rancher virtualization management and create a new one 

On the k8s-metrics page, it still display `Error: ImagePullBackOff`

```
Failed to pull image "rancher/rancher-agent:v2.6.11": rpc error: code = Unknown desc = failed to pull and unpack image
"docker.io/rancher/rancher-agent:v2.6.11": failed to resolve reference "docker.io/rancher/rancher-agent:v2.6.11": failed
to do request: Head "https://myregistry.local:5000/v2/rancher/rancher-agent/manifests/v2.6.11?ns=docker.io": dial tcp:
lookup myregistry.local on 8.8.8.8:53: read udp 192.168.2.30:36495->8.8.8.8:53: i/o timeout
```
image

I tried adding the hosts record in Harvester nodes host file but still got the same type of error 
image

**Comment by irishgordo** [comment]
@TachunLin  - I was curious, did you happen to restart the Rancher VM?

As I've noticed, it seems that the rke2-coredns settings don't persist on the Harvester VM if restarted after upgrade:

And if the Rancher VM gets restarted, the mounts don't return and docker does not autostart the registry again.
Is Docker running the registry on it at:
```
ssh vagrant@192.168.2.34 
```
pw: vagrant
And:
```
# see what docker processes are running:
sudo docker ps
# see what mounts look like:
lsblk 
# see what disk usage is:
df -h
# notice if anything suspicious is happening in journalctl, last 30 entries:
sudo journalctl -p 0..3 -n 30
```

**Comment by TachunLin** [comment]
The previous import Harvester failure would probably related to I execute the reinstall_harvester_node.sh after the Harvester provisioning failed. 

I rerun the airgapped provisioning and make sure the pxe_server -> rancher instance -> Harvester node are provision correctly in a sequence.  

Then we need to provide the bootstrap password `rancher` to login the Rancher instance 

We can follow below steps to import airgapped Harvester in Rancher 

1. First we need to set the `container-registry` in Harvester setting
    image
    
    image
    
    image

2. Double check the container-registry align with the airgapped Rancher image registry 
    ```
    {
      "Mirrors": {
        "docker.io": {
          "Endpoints": [
            "myregistry.local:5000"
          ],
          "Rewrites": null
        }
      },
      "Configs": {
        "myregistry.local:5000": {
          "Auth": null,
          "TLS": {
            "CAFile": "",
            "CertFile": "",
            "KeyFile": "",
            "InsecureSkipVerify": true
          }
        }
      },
      "Auths": null
    }
    ```

3. Then we can start importing Harvester in Rancher 
  image

4. Check the Harvester in pulling image from registry on https://192.168.2.131/dashboard/c/local/explorer#cluster-events

  image

5. Wait for a while , the import Harvester should be work
  image

6. Check the job status should be completed
  image

**Comment by TachunLin** [comment]
@irishgordo , we have some idea discussed today for the following topics:  
These could provide more detailed and complete information which can be outlined in the readme or suitable place for community user reference.  

1. We can remind user to provision the airgapped environment in the `vagrant-pxe-airgap-harvester` folder while the original `vagrant-pxe-harvester` can be used for regular non-airgapped provisioning
2. Provide the pre-requisite environment check list of the link content 
https://gist.github.com/irishgordo/dac54150863b5f51c91e48e3279587a2
3.  Provide the least available or supported Rancher and Harvester version for the airgapped solution
4.  Add check point to validate all component have installed and work as expected for example: 
    - Rancher private registry: https://myregistry.local:5000/v2/_catalog?n=500
    - Harvester cluster events: https://192.168.2.131/dashboard/c/local/explorer#cluster-events
5. Provide instruction about whe follow up action, e.g import Harvester
6. Notify use if they unexpectedly failed the provisioning of Harvester node while already complete the private registry and Rancher setup. They can't just reinstall the Harvester node but need to rebuild the provisioning from scratch again. 
7. The current supported infrastructure and stable feature of this solution, for example:
   - A fully airgapped Rancher instance with Harvester cluster 
   - Import Harvester to Rancher

**Comment by TachunLin** [comment]
@irishgordo, I try to validate the latest commit https://github.com/harvester/ipxe-examples/commit/6f2e3e0f20df01a672b3f780fef4188c1d9fe3c7 on the following scenarios, but got some failure.

1. Four nodes Harvester in airgapped (default)
2. One node Harvester in non-airgapped (set offline=false) 

  * The four nodes airgapped provisioned on an entire clean environment with sufficient hardware resources
    -  The result is It failed at **TASK [rancher : Run the equivalent of "apt-get update" as a separate step, first]** 
    ```
    task path: /home/harvester/ipxe-examples/vagrant-pxe-airgap-harvester/ansible/roles/rancher/tasks/main.yml:45
    192.168.2.34> (1, b'\r\n{"msg": "Unsupported parameters for (ansible.builtin.apt) module: update_cache_retries Supported parameters include: allow_unauthenticated, autoclean, autoremove, cache_valid_time, deb, default_release, dpkg_options, force, force_apt_get, install_recommends, only_upgrade, package, policy_rc_d, purge, state, update_cache, upgrade", "failed": true, "invocation": {"module_args": {"update_cache": true, "update_cache_retries": 10}}}\r\n', b'Shared connection to 192.168.2.34 closed.\r\n')
        <192.168.2.34> Failed to connect to the host via ssh: Shared connection to 192.168.2.34 closed.
        <192.168.2.34> ESTABLISH SSH CONNECTION FOR USER: vagrant
        <192.168.2.34> SSH: EXEC ssh -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentityFile=/home/harvester/ipxe-examples/vagrant-pxe-airgap-harvester/.vagrant/machines/rancher_box/libvirt/private_key -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="vagrant"' -o ConnectTimeout=30 -o ControlPath=/home/harvester/.ansible/cp/d8a471215c 192.168.2.34 '/bin/sh -c '"'"'rm -f -r /home/vagrant/.ansible/tmp/ansible-tmp-1683104625.1146967-28018-163338273625430/ > /dev/null 2>&1 && sleep 0'"'"''
        <192.168.2.34> (0, b'', b'')
        fatal: [arancherbox]: FAILED! => changed=false
          invocation:
            module_args:
              update_cache: true
              update_cache_retries: 10
          msg: 'Unsupported parameters for (ansible.builtin.apt) module: update_cache_retries Supported parameters include: allow_unauthenticated, autoclean, autoremove, cache_valid_time, deb, default_release, dpkg_options, force, force_apt_get, install_recommends, only_upgrade, package, policy_rc_d, purge, state, update_cache, upgrade'
    
    ```
    image


* The single nodes set with non-airgapped to test the offline=false functionality 
  - The result is it failed at **TASK [rancher : remove ip route]**

  ```
  task path: /home/davidtclin/Documents/Harvester/workspace/airgap_ipxe/0503/ipxe-examples/vagrant-pxe-airgap-harvester/ansible/roles/rancher/tasks/main.yml:17
  
    <192.168.2.34> (1, b'\r\n{"changed": true, "stdout": "", "stderr": "RTNETLINK answers: No such process", "rc": 2, "cmd": "ip route del to default via 192.168.2.254\\n", "start": "2023-05-03 02:35:11.500628", "end": "2023-05-03 02:35:11.503869", "delta": "0:00:00.003241", "failed": true, "msg": "non-zero return code", "invocation": {"module_args": {"_raw_params": "ip route del to default via 192.168.2.254\\n", "_uses_shell": true, "warn": true, "stdin_add_newline": true, "strip_empty_ends": true, "argv": null, "chdir": null, "executable": null, "creates": null, "removes": null, "stdin": null}}}\r\n', b'Shared connection to 192.168.2.34 closed.\r\n')
      <192.168.2.34> Failed to connect to the host via ssh: Shared connection to 192.168.2.34 closed
  
  <192.168.2.34> SSH: EXEC ssh -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o IdentityFile=/home/davidtclin/Documents/Harvester/workspace/airgap_ipxe/0503/ipxe-examples/vagrant-pxe-airgap-harvester/.vagrant/machines/rancher_box/libvirt/private_key -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="vagrant"' -o ConnectTimeout=30 -o ControlPath=/home/davidtclin/.ansible/cp/d8a471215c 192.168.2.34 '/bin/sh -c '"'"'rm -f -r /home/vagrant/.ansible/tmp/ansible-tmp-1683081282.6959908-22147-136103643650159/ > /dev/null 2>&1 && sleep 0'"'"''
      <192.168.2.34> (0, b'', b'')
      fatal: [arancherbox]: FAILED! => changed=true
        cmd: |-
          ip route del to default via 192.168.2.254
        delta: '0:00:00.003241'
        end: '2023-05-03 02:35:11.503869'
        invocation:
          module_args:
            _raw_params: |-
              ip route del to default via 192.168.2.254
            _uses_shell: true
            argv: null
            chdir: null
            creates: null
            executable: null
            removes: null
            stdin: null
            stdin_add_newline: true
            strip_empty_ends: true
            warn: true
        msg: non-zero return code
        rc: 2
        start: '2023-05-03 02:35:11.500628'
        stderr: 'RTNETLINK answers: No such process'
  
  
  ```
  
  image

**Comment by khushboo-rancher** [comment]
@irishgordo Can you please resolve the codeFactor errors before @TachunLin can do a final testing on this?

**Comment by irishgordo** [comment]
@khushboo-rancher  - I've updated to remove some of the CodeFactor issues :smile:  - I'm a little nervous to modify the `create-self-signed-cert.sh` 
As it's based on the [rancher2.5 installation self-signed cert](http://docs.rancher.cn/docs/rancher2.5/installation/resources/advanced/self-signed-ssl/_index/#4-%E5%A6%82%E4%BD%95%E7%94%9F%E6%88%90%E8%87%AA%E7%AD%BE%E5%90%8D%E8%AF%81%E4%B9%A6) as I've not really tested anything else as this step was based entirely on [1808's Phase 5, step 8](https://github.com/harvester/harvester/issues/1808#issuecomment-1081296179)

**Comment by khushboo-rancher** [comment]
@irishgordo the codeFactor errors are just about the concatenation of string. I think you can just eliminate the extra `"` and the errors will be resolved.
```
Khushboos-MacBook-Pro workspace % CA_CERT="dummy cert"
khushboo@Khushboos-MacBook-Pro workspace % echo -e "\033[32m ====> 2. 发现已存在CA证书，先备份"${CA_CERT}"为"${CA_CERT}"-bak，然后重新创建 \033[0m"
 ====> 2. 发现已存在CA证书，先备份dummy cert为dummy cert-bak，然后重新创建
Khushboos-MacBook-Pro workspace % echo -e "\033[32m ====> 2. 发现已存在CA证书，先备份${CA_CERT}为${CA_CERT}-bak，然后重新创建 \033[0m"
 ====> 2. 发现已存在CA证书，先备份dummy cert为dummy cert-bak，然后重新创建
```

**Comment by TachunLin** [comment]
Tested with the latest commit on with four nodes airgapped Harvester with Rancher infrastructure. 

By updating the prerequisite ansible version to 
  ```
  ansible-base >= 2.10.0 & ansible-core >= 2.11.0
  ```

And ensure the `default` network of qemu set to `Autostart`
  
  ```
  virsh net-autostart default
  ```

1. We can complete the airgapped Rancher 2.6.11 and four nodes Harvester provisioning process 

    image (1)
    image

2. And import airgapped Harvester in Rancher virtualization management 
    image

**Review by khushboo-rancher — APPROVED** [review]

File: vagrant-pxe-airgap-harvester/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+Quick Start
+-----------
+1. You can edit the `settings.yml` to change:
+    - `rancher_config.run_single_node_rancher` to `true` to enable running a single node non-airgapped Rancher instance, if you would like to airgap that Rancher instance you would set `rancher_config.run_single_node_air_gapped_rancher` to `true` as well
+    - `node_disk_size` for `rancher_config` is thin provisioned (so it won't truly take up the entire space), but if running air-gapped would need a minimum of 300G
+    - other options should be self documenting
+2. Then you can run `./setup_harvester.sh` **NOTE:** you more than likely will want to pass the flag `-c` or `--clean-ssh-known-hosts` so that when the additional configuration on each Harvester node runs after standing up Rancher old-host information is cleaned up, ex: `./setup_harvester.sh -c`, since cleaning the known_hosts file is a local file, the flag is optional, because you may want to manipulate that local file yourself instead of having this script manipulate that local file
+3.  You then can navigate to `https://<harvester_vip>:30443` to access the harvester UI.
```
Comment (khushboo-rancher) [inline vagrant-pxe-airgap-harvester/README.md:102]: I think, it is important to add the below note here.
`NOTE: by default harvester_vip is 192.168.2.131. However, it is configureable in settings.yml.`

File: vagrant-pxe-airgap-harvester/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+**Note:** if you do not desire to have an environment air-gapped it is strongly-encouraged to simply run the `vagrant-pxe-harvester` ipxe-example over this one - as this one is not *only* **more** resource intensive but will take a **substantially** long time to provision and *currently* has some known issues.
+
+**Also Note:** this set up with the `settings.yml` is currently intended to run with `rancher_config.run_single_node_rancher`, `rancher_config.run_single_node_air_gapped_rancher`, and `harvester_network_config.offline` set to `true` - there are known issues with the setup not working if those values are changed.
+
+Utilizing [Vagrant][vagrant], [KVM][kvm], and [Ansible][ansible] to create a
+ready-to-play virtual Harvester & Rancher environment for evaluation and testing
+purposes. Two Vagrant VMs will be created by default, PXE server and a
+single or multi-node Harvester respectively - with the optional vm for the Rancher instance with baked in Docker Registry, K3s & Helm installed Rancher.
+
+Known Issues
```
Comment (khushboo-rancher) [inline vagrant-pxe-airgap-harvester/README.md:16]: Can we move `Known Issues` & `troubleshooting` at the end of the Readme?

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]
This is good. Just two minor comments, then we should be able to merge this.

File: vagrant-pxe-airgap-harvester/README.md (hunk 1 above)
Comment (irishgordo) [reply vagrant-pxe-airgap-harvester/README.md]: sounds good :smile: -updated

File: vagrant-pxe-airgap-harvester/README.md (hunk 2 above)
Comment (irishgordo) [reply vagrant-pxe-airgap-harvester/README.md:62]: sounds good :smile: -updated

**Review by khushboo-rancher — APPROVED** [review]

**Pull request merged**